proposals/rich-content-graphql-api.md

```diff
@@ -0,0 +1,305 @@
+# Common GraphQL Representation of Rich Content
+
+A content-authoring user of a Magento PWA store should be able to use PageBuilder to create and edit as much of the storefront content as possible, without drastically reducing PWA performance.
+An integrator should be able to convert PageBuilder storage markup and Adobe Experience Manager UI languages, such as [Coral UI](https://helpx.adobe.com/experience-manager/6-3/sites/developing/using/reference-materials/coral-ui/coralui3/documentation.html) and [Granite UI](https://helpx.adobe.com/experience-manager/6-3/sites/developing/using/reference-materials/granite-ui/api/jcr_root/libs/granite/ui/index.html).
+
+## Vision
+
+PageBuilder builds its content in a tree of rows and columns of variable depth, whose leaf nodes are PageBuilder Blocks of various subtypes. Likewise, Coral UI and Granite UI, and most other rich content storage formats, use a tree structure of some kind, though it may not use rows, columns, and containers in the same manner. The GraphQL schema represents this with a recursively defined interface which all block types must implement. The interface consists of a short placeholder content string and enough type metadata to help the PWA query for the next layer of content. **The PWA loads PageBuilder content progressively**, by first requesting only the first two to three degrees of the content tree, then loading more content and functionality based on the metadata. As the PWA traverses PageBuilder content during a session, it caches type definitions and component implementations, and its content queries become more efficient over time. **The effect for the user is that the UI loads "from the outside in", with increasing speed as she navigates. The effect for the content author is that complex (deep) PageBuilder content will load in more "steps", incentivizing the content author to simplify layouts.**
+
+## Proposal
+
+Add an interface to the GraphQL schema representing a recursive tree of PageBuilder content nodes. Each PageBuilder block type should add a concrete type to the GraphQL schema which implements this interface and adds additional fields for any required configuration. PWAs initially query only against the interface, and can therefore only receive the metadata. In the first few content loads, the PWA must make a "followup" request with more concrete query depth. As its cache grows, it does fewer and fewer of those requests.
```

Comment:
Because the list of potential implementations is unknown by the client. Extensions that add new PageBuilder blocks could add new implementations. Those implementations would have to include a GraphQL fragment that a PWA Studio PWA could use to add to its CMS query. That can happen at build time, but it should also be possible to do so at runtime.